### PR TITLE
MSC2881: Message Attachments

### PR DESCRIPTION
**Pull request by MurzNN**
[Rendered](https://github.com/MurzNN/matrix-doc/blob/master/proposals/2881-message-attachments.md)

Depends on [MSC3051: A scalable relation format](https://github.com/matrix-org/matrix-doc/pull/3051)

Similar MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/4274

Status: Closed in favor of inline media references like this:
```
Check out my photos from FOSDEM 2025:
<img src=\"mxc://example.org/JWEIFJweifjWIeifj\" />" 
<img src=\"mxc://example.org/JWEIFJweifjWIeifk\" />" 
<img src=\"mxc://example.org/JWEIFJweifjWIeifl\" />" 

And especially look at this awesome group photo: 
<img src=\"mxc://example.org/JWEIFJweifjWIeifm\" />" 

What a nice day at FOSDEM 2025!
```
_To implement this feature, we don't need any server-side changes; this can be implemented fully on the client side (message composer to handle inserting, and renderer to display thumbnails and pop-up galleries). 

This idea is described in a separate MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/4475

Related issues:
- [Attach multiple images (files) to message](https://github.com/vector-im/element-meta/issues/685)
- [Show albums for multiple images](https://github.com/vector-im/element-meta/issues/753)
- [Message attachments (multiple photos, files, other messages for forwarding, etc)](https://github.com/matrix-org/matrix-doc/issues/2289)
- [Redesign: timeline: implement new file transfer UX](https://github.com/vector-im/element-web/issues/7565
- https://github.com/vector-im/roadmap/issues/13

Also, look up alternative and more optimal implementation: matrix-org/matrix-spec-proposals#3382

Signed-off-by: Alexey Korepov <MurzNN@gmail.com>

### CODE REVIEW

File: proposals/2881-message-attachments.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+### Fallback display
+
+For fallback display of attachments in old Matrix clients, we can attach them directly to `formatted_body` of message, here is HTML representation:
+
+```html
+<p>Here is my photos and videos from yesterday event</p>
+<div class="mx-attachments">
+  <p>Attachments:</p>
+  <ul>
+    <li><a href="https://example.com/_matrix/media/r0/download/example.com/KUAQOesGECkQTgdtedkftISg">Image 1.jpg</a></li>
```
Comment (deepbluev7) [inline proposals/2881-message-attachments.md:90]: The fallback should probably use mxc uris, as not every media repo allows downloads without authentication and it puts a lot more strain on a server, if you post in matrix.org and everyone uses the fallback instead of letting matrix.org act as a media proxy.

File: proposals/2881-message-attachments.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+This MSC does not need any changes on server side.
+
+## Potential issues
+
+The main issue is fallback display for old clients. Providing the list of links to each attachment into the formatted body is suitable workaround, and clients, which render attachments on their own, can easily remove this block via cutting `<div class="mx-attachments">` tag.
+
+## Alternatives
+
+1. An alternative can be posting media messages as separate events, as it was done earlier, and aggregating them into event via  `m.relates_to` field, but clients must do a hide of those events, when aggregating event will be added to the room (like editions), but this will be harder to implement.
```
Comment (deepbluev7) [inline proposals/2881-message-attachments.md:117]:
```suggestion
1. An alternative can be posting media messages as separate events, as it was done earlier, and aggregating them into event via  `m.relates_to` field, but clients must do a hide of those events, when aggregating event will be added to the room (like edits), but this will be harder to implement.
```

Also I disagree, that this would be any harder. Only issue may be the send order, but if you send the text message first and then the attachments after it, that should be fine. Especially if you do all the uploads before sending the first event, this should be fine. You may also add a flag to the first event, that there will be attachments, so rendering can reserve some space.

But that would give you fallbacks for free, work with existing bridges and clients and sidestep all the issues the fallback introduces. I think Whatsapp does something similar, when it groups a big drop of images into one message (just without explicit aggregations, I think).

File: proposals/2881-message-attachments.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+2. Other alternative is embedding images (and other media types) into message body via html tags, but this will make extracting and stylizing of the attachments harder.
+3. Next alternative is reuse [MSC1767: Extensible events in Matrix](https://github.com/matrix-org/matrix-doc/pull/1767) for attaching and grouping media attachments, but in current state it requires only one unique type of content per message, so we can't attach, for example, two `m.image` items into one message. Maybe, instead of separate current issue, we can extend [MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767) via converting `content` to array, to allow adding several items of same type to one message.
+
```
Comment (deepbluev7) [inline proposals/2881-message-attachments.md:120]: You may want to mention https://github.com/matrix-org/matrix-doc/pull/2530 and https://github.com/matrix-org/matrix-doc/pull/2529 as competing proposals and a justification, how this improves upon them (and potential drawbacks compared to them), like that this allows sending multiple images in one message.

File: proposals/2881-message-attachments.md (hunk 1 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:90]: In my first draft was exactly mxc urls, but I observe, that they don't work in current Element-web (and in most of other clients), so replace them to http urls. If mxc urls must work by spec, then yes - we can replace this to mxc urls.

Comment (deepbluev7) [reply proposals/2881-message-attachments.md:90]: Fair enough. mxc in `a` tags is still an open MSC: https://github.com/matrix-org/matrix-doc/pull/2398

Maybe mention that consideration in the proposal?

File: proposals/2881-message-attachments.md (hunk 3 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:121]: Yes, thanks for pointing on them, I add both to alternatives section.

File: proposals/2881-message-attachments.md (hunk 2 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:117]: Interesting idea, thanks! For fallback this way is better, but it will generate many useless events in room, instead of one grouped event. I'll try to describe this way too, but still can't decide which implementation is better... How to better describe this implementation - add section "Variant 2" in current MSC, or post a separate MSC with this implementation?

P. S. About WhatsApp - yes, do the grouping on client side, if see several events with images from one user in short time, but I didn't like this way, because of missing real grouping event and attached text description.

File: proposals/2881-message-attachments.md (hunk 1 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:90]: Yes, wasn't seen it, done now!

File: proposals/2881-message-attachments.md (hunk 2 above)
Comment (deepbluev7) [reply proposals/2881-message-attachments.md]: >  How to better describe this implementation - add section "Variant 2" in current MSC, or post a separate MSC with this implementation?

That is actually a good question. Since this MSC is fairly small, I'd just put both into this MSC by splitting the `Proposal` section into a short introduction and then 2 subheading with the 2 different variants, while naming each of those. A separate MSC is fine too, but I think as this has a clear defined scope and the implementation part is somewhat short in both cases, discussing both here should be fine. I'll let someone more experienced cast their judgement though.

File: proposals/2881-message-attachments.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+2. Other alternative is embedding images (and other media types) into message body via html tags, but this will make extracting and stylizing of the attachments harder.
+
+3. Next alternative is reuse [MSC1767: Extensible events in Matrix](https://github.com/matrix-org/matrix-doc/pull/1767) for attaching and grouping media attachments, but in current state it requires only one unique type of content per message, so we can't attach, for example, two `m.image` items into one message. Maybe, instead of separate current issue, we can extend [MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767) via converting `content` to array, to allow adding several items of same type to one message.
+
+4. There are also [MSC2530: Body field as media caption](https://github.com/matrix-org/matrix-doc/pull/2530) but it describes only text description for one media, not several media items, and very similar [MSC2529: Proposal to use existing events as captions for images](https://github.com/matrix-org/matrix-doc/pull/2529) that implement same thing, but via separate event. But if we send several medias grouped as gallery, usually one text description is enough.
+
```
Comment (deepbluev7) [inline proposals/2881-message-attachments.md:126]: I think this MSC is missing an unstable prefix: https://github.com/matrix-org/matrix-doc/pull/2371

At least `m.attachments` and `mx-attachments` would require it, imo.

File: proposals/2881-message-attachments.md (hunk 2 above)
Comment (deepbluev7) [reply proposals/2881-message-attachments.md]: I also don't see the event count as a concern. Every reaction adds an event and you don't save much bandwidth by sending everything in one event. You can also have the server aggregate everything into one event in the future, if you like, to alleviate the rendering concerns.

Comment (MurzNN) [reply proposals/2881-message-attachments.md]: I added this way as "Implementation 2: Aggregating event to group several previously sent media events" section.

File: proposals/2881-message-attachments.md (hunk 4 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:156]: Done

File: proposals/2881-message-attachments.md (hunk 2 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md]: If event count is not a concern - yes, this is better implementation that my first, so I swapped them now ;-)

File: proposals/2881-message-attachments.md (hunk 1 above)
Comment (uhoreg) [reply proposals/2881-message-attachments.md:137]: I'm surprised that an http(s) url even works (and it probably is a bug in the client if it does).  Generally clients try to avoid making requests to things other than their homeservers, to avoid leaking the user's IP address to an unknown party.

Comment (deepbluev7) [reply proposals/2881-message-attachments.md:137]: I don't think it is a bug to allow the user to click on links. It would be a different issue, if the spec allowed http(s) as a valid protocol for inline images.

Comment (uhoreg) [reply proposals/2881-message-attachments.md:137]: Ah, right, it's an `<a>`, not an ``.  Disregard my comment, then.

File: proposals/2881-message-attachments.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  "url": "mxc://example.com/KUAQOesGECkQTgdtedkftISg"
+},
+```
+And aggregating event, to send after all message attachments:
+```json
+{
+  "type": "m.room.message",
+  "content": {
+    "msgtype": "m.text",
+    "body": "Here is my photos and videos from yesterday event",
```
Comment (sumnerevans) [inline proposals/2881-message-attachments.md:39]: I think this would make it difficult to support a gallery where each of the photos has its own caption with this system. Maybe it's worth implementing both this and either #2530 or #2529 and use this for galleries, but use the the other proposal to handle captions?

File: proposals/2881-message-attachments.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      {
+            "rel_type": "m.attachment",
+            "event_id": "$id_of_previosly_send_media_event_2"
+      }
+    ]
+  }
+}
+```
+
+## Client support
```
Comment (sumnerevans) [inline proposals/2881-message-attachments.md:54]: We should probably also consider how editing should work. I'd imagine the following scenarios:

* Add caption: Add the "edit" pencil-button to pictures, and if the user clicks on that, return them to the same UI as the composer allowing them to add/remove files or edit the caption.
* Edit caption: Same flow as adding a caption.
* Remove caption: Add a "remove caption" button to the dialog if editing.
* Remove attachment: Allow attachments to be removed from the gallery with the same UI as when composing the message. One question here would be what to do with the corresponding image (or other attachment) event, probably just delete it?
* Add attachment: Allow attachments to be added to the gallery with the same UI as when composing the message. This should create a new `m.image` event and then edit the aggregating message.

File: proposals/2881-message-attachments.md (hunk 3 above)
Comment (sumnerevans) [reply proposals/2881-message-attachments.md:156]: I don't think that these proposals are necessarily mutually exclusive. There are benefits to having captions tied more directly to a single image, and then having an aggregation message that allows for multiple images to be combined into a gallery.

File: proposals/2881-message-attachments.md (hunk 5 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:39]: Message with attachments can work together with individual captions, like proposed in #2530 and #2529, because it simply refers to other event, that can have any structure. But yes - gallery, where each image have own caption, is too complex for simple chatroom, and will be not so easy to implement in UI, especially on mobiles. So I vote to keep this simple: one text message + several attachments.

File: proposals/2881-message-attachments.md (hunk 6 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:54]: As I see, editing can be represented exactly like composer interface, where user have the textarea for input message text, and area with all current attachments as tiny thumbnails, in which he can remove one of current attachments (that will remove its line from array of `m.relates_to`), add new attachment (that will upload it as new event, and refer to it in edited message).

And refer to edited message will be kept as another item of array, with `"rel_type": "m.replace"`.

Comment (sumnerevans) [reply proposals/2881-message-attachments.md:75]: On remove, should the corresponding `m.image` event also be removed? That way, clients trying to derive the state in the future will not be confused and accidentally show that image (even though it was removed from the gallery).

Comment (MurzNN) [reply proposals/2881-message-attachments.md:75]: Yes, it can (maybe even should) be redacted, like current 'redact` action on m.file events. I add this to description, thanks for pointing!

File: proposals/2881-message-attachments.md (hunk 5 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md:45]: I have added description of combining with #2530 into description of MSC.

**Review by Rebeccataylor17 — APPROVED** [review]

**Review by Rebeccataylor17 — APPROVED** [review]

File: proposals/2881-message-attachments.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  "m.relates_to": [
+    {
+      "rel_type": "m.attachment_to",
+      "event_id": "$id_of_main_message"
+    }
+  ],
+  "url": "mxc://example.com/KUAQOesGECkQTgdtedkftISg"
+},
+```      
+But this way will give harder way to render of main message event, because Matrix clients must do the search of all attached events manually in timeline, and server will can't aggregate them to main message.
```
Comment (erkinalp) [inline proposals/2881-message-attachments.md:230]:
```suggestion
But this way will give harder way to render of main message event, because Matrix clients must do the search of all attached events manually in timeline, and server will be unable to aggregate them to main message.
```

**Review by erkinalp — CHANGES_REQUESTED** [review]

**Review by erkinalp — APPROVED** [review]

**Review by erkinalp — APPROVED** [review]

**Comment by MurzNN** [comment]
The [Hummingbard](https://hummingbard.com/) project seems use "Option two" implementation from this MSC to attach multiple files/images/links to event! 🎉

File: proposals/2881-message-attachments.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
+# MSC2881: Message Attachments
+
+*This MSC is especially for media image attachments to message, but I try to make it extendable for multiple attachment types (files, videos, and in future - external URLs, links to other Matrix events, etc). So, in most of examples, I am using "image", but it means, that, instead of image, there may be another attachment type.*
+
+In the current implementation each media (file, image, video) can be sent only via a separate event to the room. But in most cases even one media is not sent alone, it must be commented via some text by the sender. So the user often wants to attach some images (one or several) directly to his text message when he is composing it,  not after or before it.
```
Comment (kevincox) [inline proposals/2881-message-attachments.md:5]:
```suggestion
In the current implementation each media (file, image, video) can be sent only via a separate event to the room. But in most cases even one media is not sent alone, it must be commented via some text by the sender. So the user often wants to attach some images (one or several) directly to their text message when composing it, not after or before it.
```

File: proposals/2881-message-attachments.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+And now the user can send images only before the message (or after it) as a separate message, but he can't attach images during the composing process to send them when the text is finished, together with the text message in one event.
```
Comment (kevincox) [inline proposals/2881-message-attachments.md:7]:
```suggestion
And now the user can send images only before the message (or after it) as a separate message, but they can't attach images during the composing process to send them when the text is finished, together with the text message in one event.
```

File: proposals/2881-message-attachments.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+With having this feature, Matrix client should allow users to attach one or multiple media (images, video, files) to message on client side, without instant sending of them to room, and send them together with text message.
+
+When user press "Send" button, Matrix client do the upload of all media, that user attached to message, as separate events to room (how it is done now), before sending message with typed text. And after sending of all attachments is finished, client send message with aggregating event, using `m.relates_to` field (from the [MSC2674: Event relationships](https://github.com/matrix-org/matrix-doc/pull/2674)), that points to all previously sent events with media, to group them into one gallery.
+
+For exclude showing those events in modern clients before grouping event added, I propose extend separate media events via adding "marker" field `is_attachment: true`, if clients got this value - they must exclude showing this media in timeline, and shows them only in gallery with grouping event.
```
Comment (kevincox) [inline proposals/2881-message-attachments.md:24]: What happens if such an event is never referenced? Does it just stay hidden forever? This is probably the best option but it is worth discussing.

Pros:
- Allows uploading the media as the user composes so that sending is quicker.
- Messages don't suddenly appear (for example) a day later higher up in the timeline.
- If the sender crashes or fails it can just restart. It isn't *required* to recollect the event IDs and reuse them. (Although reuse would be more efficient)

Cons:
- Invisible events may be an abuse vector. (DoS or covert channel)
- Upon a failed send or deleted message (if the client eagerly uploads media) the user likely doesn't realize that the media was uploaded and available in the room history. This may be a concern if the user then realized that they didn't want to send that info. (Maybe they tried to cancel the send because it was in the wrong room and would have redacted the media if they realized it was actually sent).

So basically it is nice because it emulates an atomic commit. But it is not nice because it does expose some side effects even if not committed.

File: proposals/2881-message-attachments.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+Editing interface can be represented exactly like the composer interface, where user have the textarea for input message text, and area with all current attachments as tiny thumbnails, in which he can rearrange attachments, remove one of current attachments (that will remove its line from array of `m.relates_to` and do the `redact` action on corresponding event with media in option one, or remove item from `m.attachments` in option two; and delete media file using [MSC2278](https://github.com/matrix-org/matrix-doc/blob/matthew/msc2278/proposals/2278-deleting-content.md)), add new attachment (that will upload it as new event with refer to it in edited message `m.relates_to` array in option one / added to `m.attachments` in option two).
+
+
+### Display recommendations:
+
+On the client site, attachments can be displayed as grid of clickable thumbnails, like the current `m.image` events, but with a smaller size, having fixed height, like a regular image gallery. On click, Matrix client must display media in full size, and, if possible, as a gallery with "next-previous" buttons. Also clients can implement collapse/expand action on gallery grid.
```
Comment (kevincox) [inline proposals/2881-message-attachments.md:172]: For a large number of attachments (for example "look at thses photos from my two week vacation!") many clients won't want to show all of the images upfront. Most would probably render some sort of grid with the first couple then a "37 more" link.

In order to help with cross-client consistency should we specify something about the order and how that should be handled? You do mention "rearranging" attachments earlier.

For example:
- Clients SHOULD show the attachments in the order the appear in the `m.relates_to` array.
- If the client decides to show only a subset of photos upfront it SHOULD prefer to show the ones earliest in the `m.relates_to` array.

I think it should be non-normative, but basically if you are going to only show some show the first ones rather than the last ones.

Comment (kevincox) [inline proposals/2881-message-attachments.md:172]: I don't think we need to specify "must" here. For example if the image is bigger than the users display I think most users would rather have it scaled to fit rather than needing to scroll around. I would leave this section as an example of a possible implementation.

Comment (kevincox) [inline proposals/2881-message-attachments.md:172]:
```suggestion
On the client side, attachments can be displayed as grid of clickable thumbnails, like the current `m.image` events, but with a smaller size, having fixed height, like a regular image gallery. On click, Matrix client must display media in full size, and, if possible, as a gallery with "next-previous" buttons. Also clients can implement collapse/expand action on gallery grid.
```

File: proposals/2881-message-attachments.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+Together with [MSC2675: Serverside aggregations of message relationships](https://github.com/matrix-org/matrix-doc/pull/2675) all attachments will can be even aggregated on server side.
+
+## Server support
+
+This MSC does not need any changes on server side.
+
+
+## Potential issues
+
+1. On bad connection to server Matrix client can send attachments as events with `"is_attachment": true` but not send final `m.message` event, this will lead to posting invisible media to room. This can be solved on client side via caching unsent group of events, and repeat sending when connection will be recovered.
```
Comment (kevincox) [inline proposals/2881-message-attachments.md:189]: What happens if such an event is never referenced? Does it just stay hidden forever? This is probably the best option but it is worth discussing.

Pros:
- Allows uploading the media as the user composes so that sending is quicker.
- Messages don't suddenly appear (for example) a day later higher up in the timeline.
- If the sender crashes or fails it can just restart. It isn't *required* to recollect the event IDs and reuse them. (Although reuse would be more efficient)

Cons:
- Invisible events may be an abuse vector. (DoS or covert channel)
- Upon a failed send or deleted message (if the client eagerly uploads media) the user likely doesn't realize that the media was uploaded and available in the room history. This may be a concern if the user then realizes that they didn't want to send that info. (Maybe they tried to cancel the send because it was in the wrong room and would have redacted the media if they realized it was actually sent).

So basically it is nice because it emulates an atomic commit. But it is not nice because it does expose some side effects even if not committed.

File: proposals/2881-message-attachments.md [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+2. In option one - individual media event, to which `m.message` refers, can be deleted (redacted) after. As result, `m.message` will contain relation to redacted event. In this situation Matrix clients can exclude this item from display.
```
Comment (kevincox) [inline proposals/2881-message-attachments.md:191]: I think this should be put into the MSC proper. I think it makes sense that redacting an attachment implies that the attachment is removed from the messages that include it. However I would leave it open to clients showing a "redacted" placeholder or just hiding it outright. (For example if an attachment was referenced from multiple events it may be helpful to see which events now have missing attachments).

**Review by kevincox — APPROVED** [review]

**Comment by kevincox** [comment]
This seems like a clear answer of "we will need it". It is easy to imagine and we have a real-world use case here.

File: proposals/2881-message-attachments.md [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+## Proposal
+
+### Option one "Best fallback"
+
+For solve described problem, I propose to add `m.attachment` relation type to current events, that will point to other media events in room, which must be shown as attachment to current event, and `is_attachment: true` marker field to all media, that was send to be an attachment for some message.
+
+With having this feature, Matrix client should allow users to attach one or multiple media (images, video, files) to message on client side, without instant sending of them to room, and send them together with text message.
+
+When user press "Send" button, Matrix client do the upload of all media, that user attached to message, as separate events to room (how it is done now), before sending message with typed text. And after sending of all attachments is finished, client send message with aggregating event, using `m.relates_to` field (from the [MSC2674: Event relationships](https://github.com/matrix-org/matrix-doc/pull/2674)), that points to all previously sent events with media, to group them into one gallery.
```
Comment (pavlukivan) [inline proposals/2881-message-attachments.md:22]: Can you change it to use https://github.com/matrix-org/matrix-doc/pull/3051 (and thus, m.relations instead of m.relates_to)? MSC2674 only supports a single relation per message.

File: proposals/2881-message-attachments.md (hunk 12 above)
Comment (pavlukivan) [reply proposals/2881-message-attachments.md:189]: there is already many ways to send an invisible event, that by itself shouldn't be considered an issue. The other points are still valid though

**Comment by pavlukivan** [comment]
> But so far there seem to be not enough use cases to allow multiple relations.

Off the top of my head: replying to multiple message at the same time could be done with that

File: proposals/2881-message-attachments.md (hunk 14 above)
Comment (MurzNN) [reply proposals/2881-message-attachments.md]: I added link to #3051 additionally to #2674, to keep both of them in sight.

File: proposals/2881-message-attachments.md [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+In the current implementation each media (file, image, video) can be sent only via a separate event to the room. But in most cases even one media is not sent alone, it must be commented via some text by the sender. So the user often wants to attach some images (one or several) directly to their text message when composing it, not after or before it.
+
+And now the user can send images only before the message (or after it) as a separate message, but they can't attach images during the composing process to send them when the text is finished, together with the text message in one event.
+
+On the display side, when the user sends multiple images, the problem is that each image is displayed alone, as separate event with full width in timeline, not linked to the message, and not grouped to the gallery.
+
+Messages with multiple attachments now already implemented in many messengers, for example - in Skype, Slack, VK Messenger. And Matrix, because lack of support, now have problems with bridging those messages to Matrix rooms.
+
+
+## Proposal
```
[the remark this answers is not in the document]
Comment (MurzNN) [reply proposals/2881-message-attachments.md:15]: I moved option two to separate #3382

**Comment by turt2live** [comment]
@MurzNN I'm not sure what happened here, but the diff is showing quite a few changes. I would recommend either fixing the branch, or, if preferred, closing this PR and opening a new MSC off a dedicated branch.

**Comment by richvdh** [comment]
suspect this is the problem:

image

see https://github.com/matrix-org/matrix-doc/issues/3367

**Comment by MurzNN** [comment]
@richvdh Thanks for the tip, I've switched to `main` branch, seems now all is well.

**Comment by MurzNN** [comment]
There is also a new similar MSC https://github.com/matrix-org/matrix-spec-proposals/pull/4274 but got stuck in the discussions too.

For now, I got a fresh idea of handling attachments just inline in the message body, like this:
```
Check out my photos from FOSDEM 2025:
<img src=\"mxc://example.org/JWEIFJweifjWIeifj\" />" 
<img src=\"mxc://example.org/JWEIFJweifjWIeifk\" />" 
<img src=\"mxc://example.org/JWEIFJweifjWIeifl\" />" 

And especially look at this awesome group photo: 
<img src=\"mxc://example.org/JWEIFJweifjWIeifm\" />" 

What a nice day at FOSDEM 2025!
```

Seems to implement this feature, we don't need any server-side changes; this can be implemented fully on the client side (message composer to handle inserting, and renderer to display thumbnails and popup galleries).

Additionally, we can extend the message with the list of references to inline attachments, to keep the links as dependencies (especially for encrypted messages) and be able to clean up unused (orphaned) attachments.

But, actually, I am not a Matrix client developer to move this forward with client-side implementation, so let me close this MSC with a conclusion to implement the inline attachments instead.

And if there is someone who wants to move this idea forward, please create a new MSC, and I will be happy to join and help you with the specifications details, and other discussions.

**Comment by MurzNN** [comment]
Okay, let me try with a new MSC describing a universal approach: https://github.com/matrix-org/matrix-spec-proposals/pull/4475 - let's move there!
